---
title: "How to conduct significance test for homoeologous expression ratio At/Dt?"
output: html_document
---

This was written by Guanjing Hu on Sep 29, 2014

Background
-------------
PolyCat partitions total counts for each pair of homoeologous genes into T = At + Dt + N +AD(SNPs from both found in one read), so absolute counts of At and Dt don't reflect the actual expression level of homoeologs, but only their relative relationships are reliable to infer homoeolog bias. Accordingly, all the At reads together (totalAt) and all the Dt reads together (totalDt) don't represent A and D sub-genome transcriptomes, respectively. Therefore, At and Dt reads should NOT be independently normalized by totalAt and totalDt, respectively. BUT normalization by totalT (library size) would not change their relative relationships.

To provide significance tests for the ratio At/Dt,I can think of below methods that make sense statistically, at least to me.
1. Student t-test based on At and Dt RPKMs, the resulted p-value needs to be corrected for multiple testing.
2. Differential expression using DEseq or EdgR; **importantly, the library size factors need to be estimated from T reads for the At and Dt columns.**
3. Simulation test, which I will describe below.


Simulation test using R
-------------

**Null hypothesis**: no homoeolog bias, actual At/Dt is 1, the ratio was observed by chance due to insufficent sampling.

**Alternative hypothesis**: there is homoeolog bias, the observed At/Dt ratio is different from 1.


###Simulation scheme
- Step 1. the total expression of a pair of homoeologous genes T = At + Dt + N
- Step 2. base on null hypothesis At/Dt =1, the real expressions of At and Dt should be T/2
- Step 3. randomly draw reads from T for (At+Dt) times, the sampled ratio= At'/Dt'
- Step 4. since our data have 3 replicates, we repeat step 1 to step 3 for each rep, and then calculate the average ratio.
- Step 5. run step 4 1000 times as simulation, study the distribution of sampled ratios, and estimate p-value for the observated ratio

**Note: the effects of gene length and read length were not considereed here.**

```{r}
# T is total read count, T = At + Dt + N
# At is A-specific read count
# Dt is D-specific read count
# N is common read count
# The observed ratio = At/Dt 
sti0 <- function(x=c(rep1.At, rep1.Dt, rep1.N,rep2.At, rep2.Dt, rep2.N,rep3.At, rep3.Dt, rep3.N)) {
  x<-as.numeric(x)
  # calculate the observated ratio
  ratio <-mean(c(x[1]/x[2], x[4]/x[5], x[7]/x[8]))
  # prepare to sample 1000 times  
  sampR <-rep(NA, 1000)
  # run simulation
  for(i in 1:1000) {
    # calculate total number of reads sequenced for each rep
    T <- c(sum(x[1:3]),sum(x[4:6]), sum(x[7:9]) )  
    # calculate number of diagnostic reads for each rep. At+Dt
    d <- c(sum(x[1:2]),sum(x[4:5]), sum(x[7:8]) ) 
    # draw samples from each rep
    sampRrep<-c()
    for(j in 1:3) {
      # assume At=Dt for total read counts
      pop<-c(rep("At",ceiling(T[j]/2)),rep("Dt",ceiling(T[j]/2)))
      samp<-sample(pop, d[j])
      sampRrep[j]<-length(samp[samp=="At"] )/length(samp[samp=="Dt"] )
    }
    #calculate the average ratio
    sampR[i]<-mean(sampRrep)
    }
  # Compute an empirical cumulative distribution function
  Fn <- ecdf(sampR)
  return(list(observated.ratio=ratio,  #
              confidence.interval.95=quantile(sampR,c(0.025, 0.975)),
              cumulative.p = Fn(ratio),
              p.1tail = ifelse(Fn(ratio)<0.5,Fn(ratio),1-Fn(ratio) )
              ))
  }
```

When most reads are diagnostic, low variation among reps, the observated ratio 2.4 is significantly different from null hypothesis At/Dt=1, $p.1tail<0.05.
```{r}
test1<- c(rep1.At = 66, rep1.Dt =33, rep1.N=1,
          rep2.At = 150 , rep2.Dt = 70, rep2.N =30,
          rep3.At=40, rep3.Dt=19, rep3.N=13)

sti0(test1)
```
When most reads are not diagnostic, the same observated ratio 2 is not significantly different from null hypothesis At/Dt=1, and there is no significant homoeolog bias. 
```{r}
test2<- c(rep1.At = 6, rep1.Dt =3, rep1.N=100,
          rep2.At = 14 , rep2.Dt = 7, rep2.N =300,
          rep3.At=4, rep3.Dt=2, rep3.N=130)

sti0(test2)
```

###Look at real data
I use Simon's AD1 fiber RNA-seq library to look at the relationship between N, ratio and p-value.
```{r}
# AD1 domesticated fiber 
count<-read.table("AD1_domesticated_homeo_counts.txt", sep="\t",header=TRUE)
# use 10 dpa as an example
count<-count[,grep("10DPA",names(count))]
names(count)
```
Eeach row of the data table can be given to ```sti0``` function to calculate p-value. For example, the 4th row shows significant bias:
```{r}
count[4,]
sti0(count[4,])
```
Now I try to do this with the whole table
```{r}
# get average ratio
ratio.At.Dt<-apply(count, 1, function(x){mean(c(x[1]/x[2], x[4]/x[5], x[7]/x[8]))})
# calculate p-value based simulation
p.simulation<-c()
for(gene in 1:500 ){   # test for 500 genes, otherwise it takes too long
  x<-as.numeric(count[gene,])
  if (is.na(ratio.At.Dt[gene])){p.simulation[gene]<-NA}
  else{p.simulation[gene]<-sti0(x)$p.1tail}
}
# p.simulation<-apply(count, 1, function(x){sti0(x)$p.1tail})
# do student t test, first need to normalize by library size
size<-colSums(count)  # sum of each bin
size
library.size<-c( sum(size[1:3]), sum(size[4:6]), sum(size[7:9]) )
size.factor<-library.size/mean(library.size)
factor<-rep(size.factor, each=3)
count1<-sweep(count, 2, factor, `/`)
p.ttest<-apply(count1, 1, function(x){t.test(x[c(1,4,7)],x[c(2,5,8)])$p.value})
p.ttest.1tail<-ifelse(p.ttest<=0.5,p.ttest, 1-p.ttest)
# check pvalues
hist(p.ttest.1tail)
hist(p.simulation)
plot(p.simulation, p.ttest.1tail[1:500])
```